content/blog/2023/2023-09-16-create-openfeign-http-client.md

```diff
@@ -0,0 +1,503 @@
+
+## What is Feign?
+Feign is a popular Java HTTP client library that offers several advantages and features, making it a good choice for developers building HTTP-based microservices and applications.
+
+### What is a declarative HTTP client?
+It's a way to make http requests by writing a java interface. Feign generates the actual implementation based on annotations that we provide.
+
+### Why use Feign?
+If we have a large set of APIs to call, we don't want to generate the HTTP code by hand or with code generation. It would be much easier and more maintainable to describe the API in a simple, small interface and let Feign generate the code.
+
+### Who should use Feign?
+If we are making http requests in our Java code, and don't want to write boilerplate code, or use libraries like Apache httpclient directly, Feign is a great choice.
+
+<a name="example-code" />
+{{% github "https://github.com/thombergs/code-examples/tree/master/openfeign/openfeign-client-intro" %}}
+
+## Creating a Basic Feign Client
+### Step 1: Add Feign Dependency
+    Include Feign library in the Maven pom.xml file as a dependency.
+```xml
+<dependency>
+    <groupId>io.github.openfeign</groupId>
+    <artifactId>feign-core</artifactId>
+    <version>12.5</version>
+</dependency>
+```
+### Step 2: Define Service Interface
+Define service interface. **It typically contains the method declarations annotated with Feign annotations.**
+
+We are going to declare an interface that defines signature of methods we want to call. These are just declarations. We do not implement those methods. The REST service implements those methods. These methods reperesnt the endpoints we want to call. **The method signature should include the HTTP method as well as all required data.**
+
+Let us define an interface to represent calculator service. It has simple API methods to perfom calculations like add, substract, multiply and divide.
+
+We would use Wiremock to emulate the service implementation. [WireMock](https://wiremock.org/) is a web service mocking and stubbing tool. It works by emulating a real HTTP server to which the test code can connect as if it were a real online service. It allows for HTTP response stubbing, request verification, proxy/interception, stub recording/playback, and fault injection.
+
+It is particularly useful to emulate error scenarios that are difficult to achive with real service implementation. With these emulated interactions we rest assured that when such errors occur, our client error handling logic works as expected.
+    
+Let us see our service interface.
+    
+```java
+public interface CalculatorService {
+  /**
+   * Adds two whole numbers.
+   *
+   * @param firstNumber  first whole number
+   * @param secondNumber second whole number
+   * @return sum of two numbers
+   */
+  @RequestLine("POST /operations/add?
+                      firstNumber={firstNumber}&secondNumber={secondNumber}")
```

Comment:
Please check all your code snippets for big indentations like this one ... please use only 2 spaces as indentation to conserve space ... the blog has very little space.